claat/serve.go

```diff
@@ -0,0 +1,306 @@
+// Copyright 2018 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"archive/zip"
+	"encoding/json"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+)
+
+// Global for tracking whether we've warned user about delay.
+var warned = false
+
+// cmdServe is the "claat serve ..." subcommand.
+func cmdServe() {
+	var depsDir = "bower_components"
+	var elemDir = "elements"
+	var elemFile = elemDir + "/codelab.html"
+	const codelabElem = `
+<link rel="import" href="../bower_components/google-codelab-elements/google-codelab-elements.html">
+`
+	err := os.MkdirAll(depsDir, 0755)
+	if err != nil {
+		fatalf(err.Error())
+	}
+	// Go get the dependencies.
+	err = fetchRepo(depsDir, "googlecodelabs/codelab-components")
+	if err != nil {
+		fatalf(err.Error())
+	}
+	err = os.MkdirAll(elemDir, 0755)
+	if err != nil {
+		fatalf(err.Error())
+	}
+	if _, err := os.Stat(elemFile); os.IsNotExist(err) {
+		f, err := os.Create(elemFile)
+		if err != nil {
+			fatalf(err.Error())
+		}
+		defer f.Close()
+		f.WriteString(codelabElem)
+	}
+	http.Handle("/", http.FileServer(http.Dir(".")))
+	fmt.Println("Serving on " + *addr + ", opening browser tab now...")
+	ch := make(chan error, 1)
+	go func() {
+		ch <- http.ListenAndServe(*addr, nil)
+	}()
+	openBrowser("http://" + *addr)
+	fatalf("claat serve: %v", <-ch)
+}
+
+// downloadFile will download a url to a local file. It's efficient because it will
+// write as it downloads and not load the whole file into memory.
+func downloadFile(filepath string, url string) error {
+	if !warned {
+		fmt.Println("Fetching dependencies...")
+		warned = true
+	}
+	out, err := os.Create(filepath)
+	if err != nil {
+		return err
+	}
+	defer out.Close()
+	resp, err := http.Get(url)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	_, err = io.Copy(out, resp.Body)
+	return err
+}
+
+// overrideSpec maps a given component name to desired spec, as described in fetchRepo
+// doc comments.
+var overrideSpec = map[string]string{
```

Comment:
So, why is tracking these versions directly from bower.json of dependencies not working?
I mean, the way you had it before hard coding this in a map.

Reply:
Following the recursive dep tree leads to a non-working state and I haven't managed to isolate the precise set of package/versions that are causing this problem. But if I force the versions to match our checked in bower_components, everything is fine. Conclusion: I either need to do a higher fidelity job of duplicating bower logic (e.g. pulling the "correct" minor version everywhere) or force the issue by enumerating everything. I opted for the latter to get this working. I could gradually remove the overrides until I get down to the ones that seem critical but a) that's more work on something that hopefully goes away soon, and b) I'm not sure I'll be able to tell definitively whether those version overrides are not needed (based on simple spot checking of a few codelabs).